Add pre-commit to ale

matthewgrossman opened the pull request:
Add [pre-commit](https://pre-commit.com/) as an ALE fixer

re: https://github.com/w0rp/ale/issues/1688

w0rp commented on doc/ale-precommit.txt:
```diff
@@ -0,0 +1,32 @@
+===============================================================================
+ALE pre-commit Integration                                           *ale-precommit-options*
+
```
Align the documentation tags to the right margin. `:ALEFix align_help_tags`.

matthewgrossman replied:
ah jeez, I was fighting with spaces forever on this one 😬 . This is much better

w0rp commented on doc/ale-precommit.txt:
```diff
@@ -0,0 +1,32 @@
+  Type: |List|
+  Default: `[]`
+
+  This variable can be set to pass specific hooks to the pre-commit fixer. If none
```
Wrap the line so it fits in the maximum line length. Try `gq` in visual line mode.

w0rp commented on autoload/ale/fixers/precommit.vim:
```diff
@@ -0,0 +1,29 @@
+        let l:command .= ' ' . l:options
+    endif
+
+    return {'command': ale#Escape(l:command)}
```
I just noticed, there is one mistake here. You should apply `ale#Escape()` to only `l:executable` in the command above, not the entire command. `ale#Escape` is for quoting a single part of a command, like the executable or an argument.

matthewgrossman replied:
got it, done

w0rp commented on autoload/ale/fixers/precommit.vim:
```diff
@@ -0,0 +1,29 @@
+        let l:command = l:executable . ' run --files ' . l:file_path
+    else
+        for l:hook in l:hooks
+            let l:command .= l:executable . ' run ' . l:hook . ' --files ' . l:file_path . ';'
```
You should `ale#Escape` each hook here.

matthewgrossman replied:
just to be clear, we're going to escape each `l:executable`, right? thats what the next argument seems to imply

matthewgrossman commented:
Thanks so much for the comments @w0rp! I haven't written a ton of vimscript so I appreciate the help. 

I'll be sure to get the changes merged in shortly and ping here when it's ready for another look

matthewgrossman commented on doc/ale.txt:
```diff
@@ -218,6 +218,8 @@ CONTENTS                                                         *ale-contents*
     pony..................................|ale-pony-options|
       ponyc...............................|ale-pony-ponyc|
+    pre-commit............................|ale-precommit-options|
+      pre-commit..........................|ale-precommit|
```
a single, language-agnostic fixer

w0rp replied:
You could add a `generic` language in an `ale-generic.txt` file, for any tool which works for potentially any filetype.

matthewgrossman commented on autoload/ale/fixers/precommit.vim:
```diff
@@ -0,0 +1,27 @@
+call ale#Set('precommit_hooks', [])
+call ale#Set('precommit_options', '')
+
+function! ale#fixers#precommit#Fix(buffer) abort
```
Another question @w0rp ; `pre-commit` doesn't spit the fixed file to stdout, but simply side-effects the files in place. When running the fixer locally, I essentially overwrite my own buffer with the output of stdout, which is something "useless" like

`Trim Trailing Whitespace.................................................Passed`

Is there a flag or return type I can modify the fix function to rely on the side-effecting nature of pre-commit to get this to work?

matthewgrossman replied:
```
(finished - exit code 0) ['/usr/local/bin/zsh', '-c', '''pre-commit'' run trailing-whitespace --files ''/var/folders/qz/j2l2wxfs18bbfh4vn6w00kwm0000gp/T/nvimf1gukG/13/abg_rental_provider.py'';']   
```

w0rp replied:
The fixers are normally designed to read output from a command and then apply changes to the file, typically without writing the changes back to disk before you say so. What you could do for this is define a `process_with` callback which reloads the file and returns all of the lines in it with `return getbufline(a:buffer, 1, '$')`. See `:help ale-fix`.

matthewgrossman commented:
looks like this was accomplished elsewhere!

imincik commented:
@matthewgrossman , do you know why this PR was closed without merging ? Is this feature implemented somewhere else (can't find it) ?  Thank you for working on it.

matthewgrossman commented:
ah I think you're right; I was originally doing this PR to get myself `reorder-python-imports` running as a fixer in pre-commit, but then that was added in this PR https://github.com/dense-analysis/ale/pull/2567. Once that happened, I think I no longer needed pre-commit to run as its own fixer, so I dropped the PR. Sorry for the confusion; AFAICT, pre-commit is not implemented within ALE right now

imincik commented:
Thank you very much for explanation @matthewgrossman

redbmk commented:
@matthewgrossman I still see value in precommit outside that use case. Was there anything in here specific to that pre-commit hook or were there issues you ran into getting it working?

matthewgrossman commented:
At the time I implemented it, the problem was pre-commit didn't act like "other stdin/stdout fixers". Other fixers assumed input came in to stdin and fixed-output emitted via stdout, e.g `cat myfile.txt | myfixer > original_buffer.py`. This at the time was a problem in getting it to integrate nicely with ale, though I no longer know if thats still a problem
